src/shell/filesystem_shell.rs

```diff
@@ -95,132 +95,60 @@ impl Shell for FilesystemShell {
-                                }
-
-                                let value = dir_entry_dict(filename, &metadata, &name_tag, full)?;
-                                yield ReturnSuccess::value(value);
-                            }
-                        }
-                    }
-                };
-                return Ok(stream.to_output_stream());
+                    p.push("*");
+                }
+                (p, p_tag)
             }
-        }
-
-        let entries = match glob::glob(&full_path.to_string_lossy()) {
-            Ok(files) => files,
-            Err(_) => {
-                if let Some(source) = path {
-                    return Err(ShellError::labeled_error(
-                        "Invalid pattern",
-                        "Invalid pattern",
-                        source.tag(),
-                    ));
+            None => {
+                if std::fs::read_dir(self.path()).map_or(false, |mut s| s.next().is_none()) {
+                    return Ok(OutputStream::empty());
                 } else {
-                    return Err(ShellError::untagged_runtime_error("Invalid pattern."));
+                    (PathBuf::from("./*"), context.name.clone())
                 }
             }
         };
 
-        // Enumerate the entries from the glob and add each
+        let mut paths = match glob::glob(&path.to_string_lossy()) {
+            Ok(g) => Ok(g),
+            Err(e) => Err(ShellError::labeled_error("Glob error", e.msg, &p_tag)),
+        }?
+        .peekable();
+
+        if paths.peek().is_none() {
+            return Err(ShellError::labeled_error(
+                "Invalid File or Pattern",
+                "Invalid File or Pattern",
+                &p_tag,
+            ));
+        }
+
         let stream = async_stream! {
-            for entry in entries {
+            for path in paths {
                 if ctrl_c.load(Ordering::SeqCst) {
                     break;
                 }
-                if let Ok(entry) = entry {
-                    if let Ok(metadata) = std::fs::symlink_metadata(&entry) {
-                        let mut filename = if let Ok(fname) = entry.strip_prefix(&cwd) {
-                            fname
-                        } else {
-                             Path::new(&entry)
-                        };
-
-                        if short_names {
-                            filename = if let Some(fname) = filename.file_name() {
-                                Path::new(fname)
-                            } else {
-                                let fname = filename
-                                            .file_name()
-                                            .or_else(|| Path::new(filename).file_name())
-                                            .unwrap_or(filename.as_os_str());
-                                Path::new(fname)
+                match path {
+                    Ok(p) => match std::fs::metadata(&p) {
```

Comment:
`metadata` follows symlinks, which we don't want to do right now (there's someone else working on that). Let's revert this to `symlink_metadata`.